# MC316 Final Exam – Part B: PYTHON

* You will have 1 hour and 40 minutes to finish the test.
* This part includes 10 questions. Each question is worth 15 points.
* You get extra 30 points for full answers.
* Please read the instructions carefully. 

## I. SET UP


### Import pandas library for use in scripting. 

In [126]:
import pandas as pd

### Import the datasets and name them.
For this test, you will analyze revenue and expenditure data released by Kansas Department of Administration for the fiscal year 2018. While the original data can be directly downloaded [here](http://kanview.ks.gov/DataDownload.aspx), for your convenience, the instructor had them uploaded to a server at the following urls. 

#### For Kansas State University:
1. Expenditure data: https://datajournalism.tech/wp-content/uploads/2019/05/ExpData_367_2018.csv
2. Revenue data: https://datajournalism.tech/wp-content/uploads/2019/05/RevData_367_2018.csv

#### For University of Kansas:
3. Expenditure data: https://datajournalism.tech/wp-content/uploads/2019/05/ExpData_682_2018.csv
4. Revenue data: https://datajournalism.tech/wp-content/uploads/2019/05/RevData_682_2018.csv

You will need to import all 4 datasets and name them in an easy way to call out later.

*Please refer to the [Glossary](http://kanview.ks.gov/Glossary.aspx#f) for data description.

In [127]:
expenditure = pd.read_csv("https://datajournalism.tech/wp-content/uploads/2019/05/ExpData_367_2018.csv")

In [128]:
revenue = pd.read_csv("https://datajournalism.tech/wp-content/uploads/2019/05/RevData_367_2018.csv")

In [129]:
expenditure2 = pd.read_csv("https://datajournalism.tech/wp-content/uploads/2019/05/ExpData_682_2018.csv")

In [130]:
revenue2 = pd.read_csv("https://datajournalism.tech/wp-content/uploads/2019/05/RevData_682_2018.csv")

## II. REQUIRED QUESTIONS
* Make sure you type the answers in the cells right below the codes.
* If you need to add extra cells, click the + button in the toolbar up the top. 

### Q1: What is the total revenue recorded for Kansas State University in the fiscal year of 2018? How does it compare to the total revenue recorded for University of Kansas in the same year?

In [131]:
revenue.Total.sum()

854324643.84

In [132]:
revenue2.Total.sum()

808767902.55

Answer: The total revenue recorded for K-state in the fiscal year of 2018 is higher than the total revenue recorded for University of Kansas in the same year.

In [133]:
revenue.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1910 entries, 0 to 1909
Data columns (total 22 columns):
Fiscal_Year                        1910 non-null int64
Fiscal_Quarter                     1910 non-null int64
Business_Unit                      1910 non-null int64
Business_Unit_Descr                1910 non-null object
GAAP_Fund_Type                     1910 non-null int64
GAAP_Fund_Type_Descr               1910 non-null object
Fund_Code                          1910 non-null int64
Fund_Code_Descr                    1910 non-null object
Budget_Ref                         1910 non-null int64
Budget_Ref_Descr                   1910 non-null object
Program_Code                       1910 non-null int64
Program_Code_Descr                 1910 non-null object
State_Function                     1910 non-null int64
State_Function_Descr               1910 non-null object
Primary_Account_Code               1910 non-null int64
Primary_Account_Code_Descr         1910 non-null object
Interm

### Q2: What is the total expense recorded for Kansas State University in the fiscal year of 2018? How does it compare to the total expense recorded for University of Kansas in the same year?

In [134]:
expenditure.Total.sum()

1018584373.02

In [135]:
expenditure2.Total.sum()

990238875.6099999

Answer: The total expense recorded for K-STATE is also higher than the University of Kansas in the same year.

### Q3: Where did the two universities' primary revenues come from? 
*Refer to the description [here](http://kanview.ks.gov/Revenues/AcctCodeRevenues.aspx) to write your answer.

In [136]:
revenue.Primary_Account_Code_Descr.value_counts()

AGENCY EARNINGS                   637
REV FROM USE OF MONEY & PROP      354
REVENUE TRANSFERS                 341
NON REVENUE RECEIPTS              259
OTHER REVENUE                     160
GIFT DONATION FEDERAL GRANTS      152
TAXES                               7
Name: Primary_Account_Code_Descr, dtype: int64

<b>Answer:</b> They are from agency earnings, use of money & prop, revenue transfers, revenue receipts, other revenue gift donation federal grants and taxes.

In [137]:
revenue2.Primary_Account_Code_Descr.value_counts()

AGENCY EARNINGS                   376
REV FROM USE OF MONEY & PROP      251
NON REVENUE RECEIPTS              157
REVENUE TRANSFERS                 153
OTHER REVENUE                      57
TAXES                              45
GIFT DONATION FEDERAL GRANTS       38
Name: Primary_Account_Code_Descr, dtype: int64

### Q4: What was the biggest source of revenues for Kansas State University? Compare to the one of University of Kansas.

Remove hashtag and run the following code to suppress scientific notation:

In [138]:
pd.set_option('display.float_format', lambda x: '%.3f' % x) 

In [139]:
revenue.groupby("Primary_Account_Code_Descr").Total.sum()

Primary_Account_Code_Descr
AGENCY EARNINGS                  412565112.820
GIFT DONATION FEDERAL GRANTS     278778848.030
NON REVENUE RECEIPTS              71947564.050
OTHER REVENUE                      8432030.160
REV FROM USE OF MONEY & PROP      34139318.040
REVENUE TRANSFERS                 48458616.490
TAXES                                 3154.250
Name: Total, dtype: float64

In [140]:
revenue2.groupby("Primary_Account_Code_Descr").Total.sum()

Primary_Account_Code_Descr
AGENCY EARNINGS                  408409337.320
GIFT DONATION FEDERAL GRANTS     193836668.320
NON REVENUE RECEIPTS              69753247.670
OTHER REVENUE                     68241930.500
REV FROM USE OF MONEY & PROP      40950779.540
REVENUE TRANSFERS                 26013855.400
TAXES                              1562083.800
Name: Total, dtype: float64

<b>Answer:</b>The biggest source of revenues for K-state University is agency earnings, and the amount of K-state's agency earnings are bigger than the university of Kansas.

### Q5: What accounts contributed to Kansas State University's biggest source of revenues?
If you run into a situation where a filter function returns 0 results, you may want to check if there are leading and trailing spaces in the column. To clean the leading and trailing spaces, you will need to remove the hashtag, revise the names, and run the following code:

In [141]:
revenue["Primary_Account_Code_Descr"] = revenue["Primary_Account_Code_Descr"].str.strip()

After cleaning the spaces, you can start filtering your datasets to find the answer for Q5.

In [142]:
data = revenue[revenue.Primary_Account_Code_Descr == "AGENCY EARNINGS"]

In [143]:
data.head()

Fiscal_Year  Fiscal_Quarter  Business_Unit  \
2          2018               1            367   
7          2018               1            367   
8          2018               1            367   
12         2018               1            367   
13         2018               1            367   

               Business_Unit_Descr  GAAP_Fund_Type  \
2   Kansas State University                     12   
7   Kansas State University                     12   
8   Kansas State University                     12   
12  Kansas State University                     12   
13  Kansas State University                     12   

              GAAP_Fund_Type_Descr  Fund_Code                 Fund_Code_Descr  \
2   Component Unit                       2062  GENERAL FF                       
7   Component Unit                       2062  GENERAL FF                       
8   Component Unit                       2129  GENERAL FF                       
12  Component Unit                       2154  KAN-GROW ENGINEERING FUND-KSU    
13  Component Unit                       2263  FERTILIZER RESEARCH FD           

    Budget_Ref                                   Budget_Ref_Descr  ...  \
2         2000  GENERAL FF                                    ...  ...   
7         2000  GENERAL FF                                    ...  ...   
8         5500  GENERAL FF                                    ...  ...   
12        2154  KAN-GROW ENGINEERING FUND-KSU                 ...  ...   
13        1150  FERTILIZER RESEARCH FD                        ...  ...   

    State_Function                               State_Function_Descr  \
2               26  Universities                                  ...   
7               26  Universities                                  ...   
8               26  Universities                                  ...   
12              26  Universities                                  ...   
13              26  Universities                                  ...   

    Primary_Account_Code Primary_Account_Code_Descr  \
2                   4200            AGENCY EARNINGS   
7                   4200            AGENCY EARNINGS   
8                   4200            AGENCY EARNINGS   
12                  4200            AGENCY EARNINGS   
13                  4200            AGENCY EARNINGS   

    Intermediate_Account_Code Intermediate_Account_Code_Descr  \
2                       42000  SERVICE CHARGES                  
7                       42000  SERVICE CHARGES                  
8                       42000  SERVICE CHARGES                  
12                      42000  SERVICE CHARGES                  
13                      42000  SERVICE CHARGES                  

    Detail_Account_Code       Detail_Account_Code_Descr        Total  \
2                420500  EDUCATION AND LIBRARIES        91090183.660   
7                420500  EDUCATION AND LIBRARIES         3260669.860   
8                420500  EDUCATION AND LIBRARIES         7959879.110   
12               420200  TECHNICAL AND SKILLED SERVICES  3150000.000   
13               420100  AUDITS AND INSPECTIONS            56932.070   

   Unnamed: 21  
2          nan  
7          nan  
8          nan  
12         nan  
13         nan  

[5 rows x 22 columns]

<b>Answer:</b>420500 account countribute the biggest source of revenues.

### Q6: Where did the money go? (Answer for both universities)
*Refer to the description [here] (http://kanview.ks.gov/Expenditures/AcctCodeExpenditures.aspx) to write your answer.

In [144]:
expenditure.Primary_Account_Code_Descr.value_counts()

CONTRACTUAL SERVICES              10561
SALARIES AND WAGES                 7780
COMMODITIES                        3740
CAPITAL OUTLAY                     1173
GRANTS CLAIMS SHARED REVENUE        534
NONEXPENSE ITEMS                    129
EXPENSE TRANSFERS                   106
DEBT SERVICE                         71
Name: Primary_Account_Code_Descr, dtype: int64

In [145]:
expenditure2.Primary_Account_Code_Descr.value_counts()

CONTRACTUAL SERVICES              4322
SALARIES AND WAGES                3623
COMMODITIES                       1605
CAPITAL OUTLAY                     737
NONEXPENSE ITEMS                    81
GRANTS CLAIMS SHARED REVENUE        75
EXPENSE TRANSFERS                   66
DEBT SERVICE                        44
Name: Primary_Account_Code_Descr, dtype: int64

<b>Answer:</b> The money goes to cintractual services, salaries and wages, commodities, capital outlay, grants claims shared revenue, debit service, expense transfers, nonexpense items.

### Q7: What costs two universities the most? 

In [146]:
expenditure.groupby("Primary_Account_Code_Descr").Total.sum()

Primary_Account_Code_Descr
CAPITAL OUTLAY                    64350550.130
COMMODITIES                       42670908.840
CONTRACTUAL SERVICES             128541598.280
DEBT SERVICE                      32442445.100
EXPENSE TRANSFERS                 37130576.230
GRANTS CLAIMS SHARED REVENUE     106551248.640
NONEXPENSE ITEMS                 130529992.790
SALARIES AND WAGES               476367053.010
Name: Total, dtype: float64

In [147]:
expenditure2.groupby("Primary_Account_Code_Descr").Total.sum()

Primary_Account_Code_Descr
CAPITAL OUTLAY                    70676545.250
COMMODITIES                       24225857.120
CONTRACTUAL SERVICES             139630507.780
DEBT SERVICE                      25309742.030
EXPENSE TRANSFERS                 25815417.520
GRANTS CLAIMS SHARED REVENUE      71439705.360
NONEXPENSE ITEMS                 161145084.400
SALARIES AND WAGES               471996016.150
Name: Total, dtype: float64

<b>Answer:</b> Salaries and wages costs two universities the most.

### Q8: Calculate the percentage of this cost over each university's total expense.

In [148]:
476367053.010/1018584373.02

0.4676755952946929

In [149]:
471996016.150/990238875.6099999

0.47664864284311637

<b>Answer:</b> The persentage of the Kansas state university is 46.76%, and the persentafe of the University of Kansas is 47.66%.

### Q9: Breakdown the cost you found in Q7 for both universities.
Again, if you run into a situation where a filter function returns 0 results, you may want to check if there are leading and trailing spaces in the column. To clean the leading and trailing spaces, you will need to remove the hashtag, revise the names, and run the following code:

In [150]:
expenditure["Primary_Account_Code_Descr"] = expenditure["Primary_Account_Code_Descr"].str.strip()

In [151]:
data2 = expenditure[expenditure.Primary_Account_Code_Descr == "SALARIES AND WAGES"]

In [154]:
expenditure.Detail_Account_Code_Descr.value_counts()

WORKERS COMPENSATION              576
LEAVE PAYMENT ASSESSMENTS         575
UNEMPLOYMENT COMPENSATION         542
MEDICARE                          541
OASDI                             534
UNCLASSIFIED REGULAR              525
UNCLASS HOLIDAY INCL IN BASE      510
GROUP HEALTH HOSPITALIZATION      505
EDUCATIONAL INST EE RETIRE SYS    474
PARKING COMPENSATION REDUCTION    468
STUDENT EMPLOYEES HOURLY          434
PUBLIC EE RETIREMENT SYSTEM       384
PRO SCIENTIFIC SUPPLY OTHER       366
ED INST EE GRP LIFE DISABILITY    364
OTHER FEES                        347
INTERGOV OTHER COMM NOT OITS      328
INTERGOV LCL COMM SVC NON OITS    325
DATA PROCESSING SUPPLIES          325
STATIONERY AND OFFICE SUPPLIES    313
JOB RELATED TRAIN & CONF REG      312
INTERGOV LD COMM SVC NOT OITS     306
MEALS AND LODGING OUT STATE       300
OTHER PROFESSIONAL FEES           299
PRINT BIND STATE FACILITY ONLY    293
AIR RAIL & BUS FARE OUT STATE     284
UNCLASSIFIED TEMPORARY            275
REPAIR AND S

In [163]:
data2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7780 entries, 16 to 24089
Data columns (total 22 columns):
Fiscal_Year                        7780 non-null int64
Fiscal_Quarter                     7780 non-null int64
Business_Unit                      7780 non-null int64
Business_Unit_Descr                7780 non-null object
GAAP_Fund_Type                     7780 non-null int64
GAAP_Fund_Type_Descr               7780 non-null object
Fund_Code                          7780 non-null int64
Fund_Code_Descr                    7780 non-null object
Budget_Ref                         7780 non-null int64
Budget_Ref_Descr                   7780 non-null object
Program_Code                       7780 non-null int64
Program_Code_Descr                 7780 non-null object
State_Function                     7780 non-null int64
State_Function_Descr               7780 non-null object
Primary_Account_Code               7780 non-null int64
Primary_Account_Code_Descr         7780 non-null object
Inte

In [ ]:
expenditure2["Primary_Account_Code_Descr"] = expenditure2["Primary_Account_Code_Descr"].str.strip()

In [ ]:
data3 = expenditure2[expenditure2.Primary_Account_Code_Descr == "SALARIES AND WAGES"]

In [155]:
expenditure2.Detail_Account_Code_Descr.value_counts()

OTHER FEES                        235
WORKERS COMPENSATION              225
LEAVE PAYMENT ASSESSMENTS         225
MEDICARE                          219
OASDI                             219
UNEMPLOYMENT COMPENSATION         219
UNCLASSIFIED REGULAR              213
EDUCATIONAL INST EE RETIRE SYS    208
GROUP HEALTH HOSPITALIZATION      207
UNCLASS HOLIDAY INCL IN BASE      197
PARKING COMPENSATION REDUCTION    179
STUDENT EMPLOYEES HOURLY          160
ED INST EE GRP LIFE DISABILITY    156
PUBLIC EE RETIREMENT SYSTEM       154
PRO SCIENTIFIC SUPPLY OTHER       134
STATIONERY AND OFFICE SUPPLIES    131
CLASSIFIED REGULAR                126
DATA PROCESSING SUPPLIES          126
OTHER PROFESSIONAL FEES           121
PRINTING AND COPYING              115
BLDGS AND GROUNDS MATERIALS       113
CLASS HOLIDAY INCL IN BASE        112
MEALS AND LODGING OUT STATE       112
JOB RELATED TRAIN & CONF REG      110
UNCLASS HOL NOT INCL IN BASE      109
NON SUBSISTENCE OUT STATE         105
UNCLASSIFIED

In [164]:
data3.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3623 entries, 1 to 10518
Data columns (total 22 columns):
Fiscal_Year                        3623 non-null int64
Fiscal_Quarter                     3623 non-null int64
Business_Unit                      3623 non-null int64
Business_Unit_Descr                3623 non-null object
GAAP_Fund_Type                     3623 non-null int64
GAAP_Fund_Type_Descr               3623 non-null object
Fund_Code                          3623 non-null int64
Fund_Code_Descr                    3623 non-null object
Budget_Ref                         3623 non-null int64
Budget_Ref_Descr                   3623 non-null object
Program_Code                       3623 non-null int64
Program_Code_Descr                 3623 non-null object
State_Function                     3623 non-null int64
State_Function_Descr               3623 non-null object
Primary_Account_Code               3623 non-null int64
Primary_Account_Code_Descr         3623 non-null object
Inter

<b>Answer:</b> The cost breakdown should be related to salaries and wages.

### Q10: Compare the breakdown costs from Q9 side by side.
When you merge the two subsets, make sure to add the attribute <b>how ="outer"</b> so that you can keep every item from both sets.
*For example: merge = pd.merge(df1, df2, how="outer", on="columnname")

In [162]:
merged = pd.merge(data2, data3, on="Primary_Account_Code_Descr")

In [161]:
merged.head()

Fiscal_Year_x  Fiscal_Quarter_x  Business_Unit_x  \
0           2018                 1              367   
1           2018                 1              367   
2           2018                 1              367   
3           2018                 1              367   
4           2018                 1              367   

            Business_Unit_Descr_x  GAAP_Fund_Type_x  \
0  Kansas State University                       12   
1  Kansas State University                       12   
2  Kansas State University                       12   
3  Kansas State University                       12   
4  Kansas State University                       12   

           GAAP_Fund_Type_Descr_x  Fund_Code_x  \
0  Component Unit                         3047   
1  Component Unit                         3047   
2  Component Unit                         3047   
3  Component Unit                         3047   
4  Component Unit                         3047   

                Fund_Code_Descr_x  Budget_Ref_x  \
0  SMITH-LEVER SPEC PRG GRNT FDF           1330   
1  SMITH-LEVER SPEC PRG GRNT FDF           1330   
2  SMITH-LEVER SPEC PRG GRNT FDF           1330   
3  SMITH-LEVER SPEC PRG GRNT FDF           1330   
4  SMITH-LEVER SPEC PRG GRNT FDF           1330   

                                  Budget_Ref_Descr_x  ...  \
0  SMITH-LEVER SPEC PRG GRNT FDF                 ...  ...   
1  SMITH-LEVER SPEC PRG GRNT FDF                 ...  ...   
2  SMITH-LEVER SPEC PRG GRNT FDF                 ...  ...   
3  SMITH-LEVER SPEC PRG GRNT FDF                 ...  ...   
4  SMITH-LEVER SPEC PRG GRNT FDF                 ...  ...   

             Program_Code_Descr_y State_Function_y  \
0  Academic Support                              3   
1  Academic Support                              3   
2  Academic Support                              3   
3  Academic Support                              3   
4  Academic Support                              3   

                              State_Function_Descr_y Primary_Account_Code_y  \
0  Education                                     ...                   5100   
1  Education                                     ...                   5100   
2  Education                                     ...                   5100   
3  Education                                     ...                   5100   
4  Education                                     ...                   5100   

   Intermediate_Account_Code_y Intermediate_Account_Code_Descr_y  \
0                        51000    SALARIES AND WAGES               
1                        51000    SALARIES AND WAGES               
2                        51000    SALARIES AND WAGES               
3                        51000    SALARIES AND WAGES               
4                        51000    SALARIES AND WAGES               

   Detail_Account_Code_y     Detail_Account_Code_Descr_y  Total_y  \
0                 511100  UNCLASSIFIED REGULAR           2908.660   
1                 517600  LEAVE PAYMENT ASSESSMENTS        21.810   
2                 517800  PARKING COMPENSATION REDUCTION    0.920   
3                 518300  EDUCATIONAL INST EE RETIRE SYS  247.230   
4                 519101  MEDICARE                         41.160   

  Unnamed: 21_y  
0           nan  
1           nan  
2           nan  
3           nan  
4           nan  

[5 rows x 43 columns]

<b>Answer:</b>

## III. BONUS QUESTIONS
* Each question's worth 15 points.


### B1: Which university made a profit in the fiscal year of 2018?

In [166]:
854324643.84/1018584373.02

0.8387372381406301

In [167]:
808767902.55/990238875.6099999

0.8167402052881316

<b>Answer:</b> Kansas state university.

### B2: Which software program do you like the most in this course? Why?

<b>Answer:</b> I prefer using tableau and google sheet. Because it is convenience and easily to use to creat charts and clearn data.

### END OF THE TEST